File: .opencode/skills/neon-translator-runtime/build.test.mjs
#!/usr/bin/env node
import { readFileSync, existsSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { execSync, spawnSync } from 'node:child_process';

const __dirname = dirname(fileURLToPath(import.meta.url));
const GAME_DIR = join(__dirname, '..', '..', '..');
const CSC = 'C:/Windows/Microsoft.NET/Framework64/v4.0.30319/csc.exe';

let passed = 0;
let failed = 0;
function assert(cond, msg) {
  if (cond) { passed++; process.stdout.write('.'); }
  else { failed++; process.stdout.write('F'); console.error('\n  FAIL:', msg); }
}

async function main() {
  console.log('=== NeonTranslatorRuntime — Build Tests ===\n');

  console.log('1. Source files');
  for (const f of ['NativeMethods.cs','TranslationLoader.cs','MethodPatcher.cs','TranslatorPlugin.cs'])
    assert(existsSync(join(__dirname, 'source', f)), f + ' exists');

  console.log('\n2. Build script');
  assert(existsSync(join(__dirname, 'build.mjs')), 'build.mjs exists');

  console.log('\n3. Compilation');
  try {
    execSync('node "' + join(__dirname, 'build.mjs') + '"', { stdio: 'pipe', timeout: 120000 });
  } catch(e) { assert(false, 'build: ' + (e.stderr||'').substring(0,100)); }

  const dllPath = join(GAME_DIR, 'output', 'runtime', 'NeonTranslatorRuntime.dll');
  assert(existsSync(dllPath), 'DLL created');
  const dllSize = readFileSync(dllPath).length;
  assert(dllSize > 1000, 'DLL size: ' + dllSize + ' bytes');

  console.log('\n4. DLL content verification');
  const content = readFileSync(dllPath).toString('latin1');
  assert(content.includes('NeonTranslator'), 'NeonTranslator namespace');
  assert(content.includes('TranslatorPlugin'), 'TranslatorPlugin');
  assert(content.includes('MethodPatcher'), 'MethodPatcher');
  assert(content.includes('TranslationLoader'), 'TranslationLoader');
  assert(content.includes('NativeMethods'), 'NativeMethods');
  assert(content.includes('VirtualProtect'), 'VirtualProtect');

  console.log('\n5. TranslationLoader unit test');
  const testFile = join(__dirname, 'test_data', 'test.ndjson');
  assert(existsSync(testFile), 'test.ndjson exists');
  const testContent = readFileSync(testFile, 'utf-8');
  assert(testContent.includes('Resolution Scaling'), 'test data present');
  assert(testContent.includes('Масштабирование разрешения'), 'test translation present');

  // Compile and run a mini test via csc
  console.log('\n6. Quick compile check');
  const testProg = join(__dirname, 'test_data', 'test_loader.cs');
  if (existsSync(testProg)) {
    const testOut = join(GAME_DIR, 'output', 'runtime', 'test_loader.exe');
    const MD = join(GAME_DIR, 'Third Crisis Neon Nights_Data', 'Managed');
    const srcDir = join(__dirname, 'source');
    const args = [
      '/target:exe', '/out:' + testOut, '/nologo',
      '/r:' + join(MD, 'netstandard.dll'),
      '/r:' + join(MD, 'UnityEngine.dll'),
      '/r:' + join(MD, 'UnityEngine.CoreModule.dll'),
      testProg,
      join(srcDir, 'TranslationLoader.cs'),
      join(srcDir, 'NativeMethods.cs'),
    ];
    const r = spawnSync(CSC, args, { stdio: 'pipe', timeout: 60000, encoding: 'utf8', cwd: GAME_DIR });
    assert(r.status === 0, 'test_loader compiled: ' + (r.stdout||'').substring(0,100));
    if (r.status === 0) {
      try {
        const run = spawnSync(testOut, [testFile], { stdio: 'pipe', timeout: 10000, encoding: 'utf8', cwd: GAME_DIR });
        assert(run.status === 0, 'test_loader ran OK');
      } catch(e) { assert(false, 'test_loader run failed'); }
    }
  } else {
    assert(true, 'test_loader.cs not found (skipped)');
  }

  console.log('\n\n=== Result:', passed, 'passed,', failed, 'failed ===');
  process.exit(failed > 0 ? 1 : 0);
}
main().catch(e => { console.error('FATAL:', e); process.exit(1); });

File: .opencode/skills/build-translator/build.test.py
import os, sys, subprocess
from pathlib import Path

SKILL_DIR = Path(__file__).parent
GAME_DIR = SKILL_DIR.parent.parent.parent
CSC = r'C:\Windows\Microsoft.NET\Framework64\v4.0.30319\csc.exe'

passed = 0
failed = 0

def assert_eq(cond, msg):
    global passed, failed
    if cond:
        passed += 1
        print('.', end='', flush=True)
    else:
        failed += 1
        print('F', end='', flush=True)
        print(f'\n  FAIL: {msg}')

def main():
    global passed, failed

    print('=== NeonTranslatorRuntime — Build Tests ===\n')

    print('1. Source files')
    for f in ['NativeMethods.cs', 'TranslationLoader.cs', 'MethodPatcher.cs', 'TranslatorPlugin.cs', 'NeonLateUpdate.cs']:
        p = SKILL_DIR / 'source' / f
        assert_eq(p.exists(), f'{f} exists')

    print()
    print('2. Build script')
    assert_eq((SKILL_DIR / 'build.py').exists(), 'build.py exists')

    print()
    print('3. Compilation')
    build_py = SKILL_DIR / 'build.py'
    try:
        result = subprocess.run([sys.executable, str(build_py)], capture_output=True, text=True, timeout=120, cwd=str(GAME_DIR))
        assert_eq(result.returncode == 0, f'build: {result.stdout[:200]}')
    except Exception as e:
        assert_eq(False, f'build exception: {e}')

    dll_path = GAME_DIR / 'runtime' / 'NeonTranslatorRuntime.dll'
    assert_eq(dll_path.exists(), 'DLL created')
    dll_size = dll_path.stat().st_size if dll_path.exists() else 0
    assert_eq(dll_size > 1000, f'DLL size: {dll_size} bytes')

    print()
    print('4. DLL content verification')
    if dll_path.exists():
        content = dll_path.read_bytes()
        for term in ['NeonTranslator', 'TranslatorPlugin', 'MethodPatcher', 'TranslationLoader', 'NativeMethods', 'VirtualProtect']:
            assert_eq(term.encode() in content, f'{term} present')

    print()
    print('5. TranslationLoader unit test')
    test_file = SKILL_DIR / 'test_data' / 'test.ndjson'
    assert_eq(test_file.exists(), 'test.ndjson exists')
    if test_file.exists():
        test_content = test_file.read_text(encoding='utf-8')
        assert_eq('Resolution Scaling' in test_content, 'test data present')
        assert_eq('Масштабирование разрешения' in test_content, 'test translation present')

    print()
    print('6. Quick compile check')
    test_prog = SKILL_DIR / 'test_data' / 'test_loader.cs'
    if test_prog.exists():
        out_dir = GAME_DIR / 'runtime'
        out_dir.mkdir(parents=True, exist_ok=True)
        test_out = out_dir / 'test_loader.exe'
        md = GAME_DIR / 'Third Crisis Neon Nights_Data' / 'Managed'
        src_dir = SKILL_DIR / 'source'
        args = [
            CSC, '/target:exe', f'/out:{test_out}', '/nologo',
            f'/r:{md / "netstandard.dll"}',
            f'/r:{md / "UnityEngine.dll"}',
            f'/r:{md / "UnityEngine.CoreModule.dll"}',
            str(test_prog),
            str(src_dir / 'TranslationLoader.cs'),
            str(src_dir / 'NativeMethods.cs'),
        ]
        try:
            r = subprocess.run(args, capture_output=True, text=True, timeout=60, cwd=str(GAME_DIR))
            assert_eq(r.returncode == 0, f'test_loader compiled: {r.stdout[:200]}')
            if r.returncode == 0:
                try:
                    run = subprocess.run([str(test_out), str(test_file)], capture_output=True, text=True, timeout=10, cwd=str(GAME_DIR))
                    assert_eq(run.returncode == 0, 'test_loader ran OK')
                except:
                    assert_eq(False, 'test_loader run failed')
        except:
            assert_eq(False, 'test_loader compile exception')
    else:
        assert_eq(True, 'test_loader.cs not found (skipped)')

    print()
    print(f'\n\n=== Result: {passed} passed, {failed} failed ===')
    sys.exit(1 if failed > 0 else 0)

if __name__ == '__main__':
    main()
